### PR TITLE
Create Terraform Domains demo

### PR DESCRIPTION
**Pull request by ghost**
This is a demo for showcasing the Terraform Provider and using it to register and manage domains.

Belongs to https://github.com/dnsimple/dnsimple-external-integrations/issues/9 and https://github.com/dnsimple/dnsimple-marketing/issues/619.

### CODE REVIEW

File: terraform/domains/README.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# DNSimple Terraform Domains Demo
+
+This is a demo of using the DNSimple Terraform Provider to manage your DNSimple domains with the simplicity and automation of Terraform and Infrastructure-as-Code. This demo showcases how straightforward and easy it is to register, manage, and synchronise your entire domain collection, whether one or one thousand. Many of you will have lots of domains, for various purposes: regional domains, development and infrastructure domains, internal company domains, typosquatting domains, associated brands or marketing domains, and more. With the Terraform provider, you can use regular Terraform syntax and resources, and leverage all of Terraform's features and advantages, to ensure your domains are protected and high quality without any error-prone, tedious, or confusing manual processes or external tools. You can learn more on our [blog](https://blog.dnsimple.com/2023/06/terraform-domain-registrations/).
```
Comment (itsalyse) [inline terraform/domains/README.md:3]:
```suggestion
This is a demo of using the DNSimple Terraform Provider to manage your DNSimple domains with the simplicity and automation of Terraform and Infrastructure-as-Code. This demo showcases how straightforward and easy it is to register, manage, and synchronise your entire domain collection, whether one or one thousand. Many of you will have lots of domains for various purposes: regional domains, development and infrastructure domains, internal company domains, typosquatting domains, associated brand or marketing domains, and more. With the Terraform provider, you can use regular Terraform syntax and resources, and leverage all of Terraform's features and advantages, to ensure your domains are protected and high quality — without any error-prone, tedious, or confusing manual processes or external tools. You can learn more on our [blog](https://blog.dnsimple.com/2023/06/terraform-domain-registrations/).
```

File: terraform/domains/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+What this demo does:
+
+- Creates a contact to be used for all of your domains.
+- Defines a list of domains in [domains.json](./domains.json), which will be provided to Terraform.
+- Registers and synchronises all these domains using the `dnsimple_registered_domain` Terraform resource.
```
Comment (itsalyse) [inline terraform/domains/README.md:9]:
```suggestion
- Registers and synchronizes all these domains using the `dnsimple_registered_domain` Terraform resource.
```

File: terraform/domains/README.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+With this simple combo, you can:
+
+- Register more domains simply by adding a line to `domains.json`.
+- Detect drift at any domain, which could be caused by unintentional or suspicious activity.
+- Keep your entire set of domain properties in sync and managed.
```
Comment (itsalyse) [inline terraform/domains/README.md:15]:
```suggestion
- Keep your entire set of domain properties in-sync and managed.
```

File: terraform/domains/README.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+- Update contacts for all domains with just one command.
+
+You can build on top of this simple core usage by leveraging many other DNSimple features, like DNS, SSL certificates, email forwarding, and more, using the same Terraform script and approach, with the same simplicity and automation at scale. There are many demos for these within this repo; check them out [here](../).
+
+## Prerequisites
+
+To run this demo, you will need an API token for your DNSimple account must be generated and stored in your environment variables. See [here](https://support.dnsimple.com/articles/api-access-token/) for more details.
```
Comment (itsalyse) [inline terraform/domains/README.md:22]:
```suggestion
To run this demo, you will need an API token for your DNSimple account that must be generated and stored in your environment variables. See [here](https://support.dnsimple.com/articles/api-access-token/) for more details.
```

File: terraform/domains/README.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+- Adding a new domain to `domains.json`.
+- Changing the `whois_privacy_enabled` attribute for all domains, and seeing if these are reflected when you do a [WHOIS lookup](https://dnsimple.com/whois).
+- Updating the contact, and then doing a [WHOIS lookup](https://dnsimple.com/whois).
+- Extending `domains.json` from an array of strings to an object, to represent different configuration for different domains. This will require some additional code in `main.tf` too.
+- Setting up a [URL record](https://support.dnsimple.com/articles/url-record/) for all typosquatting to redirect to the official domain. You may find our [DNS demo](../dns) helpful.
+
+## Notes
+
+Your domains aren't deleted when you destroy the Terraform resources, for safety reasons. This means that those domains will remain registered to you even after you complete this demo and clean up the Terraform resources.
```
Comment (itsalyse) [inline terraform/domains/README.md:53]:
```suggestion
For safety reasons, your domains aren't deleted when you destroy the Terraform resources. This means those domains will remain registered to you even after you complete this demo and clean up the Terraform resources.
```

**Review by itsalyse — APPROVED** [review]

File: terraform/domains/main.tf [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  city = "San Francisco"
+  state_province = "California"
+  postal_code = "90210"
+  country = "US"
+  phone = "+18002203458"
+  email = "domains@trusty.com"
+}
+
+resource "dnsimple_registered_domain" "domain" {
+  for_each = toset(jsondecode(file("domains.json")))
```
Comment (DXTimer) [inline terraform/domains/main.tf:29]: The `domains.json` may not be a practical approach for domain registrations as some TLDs may require extended attributes, also some TLDs may not support transfer lock, whois privacy, or DNSSEC.

File: terraform/domains/main.tf [hunk 7]
```diff
@@ -0,0 +1,37 @@
+terraform {
+  required_providers {
+    dnsimple = {
+      source  = "dnsimple/dnsimple"
+      version = "1.2.0"
+    }
+  }
+}
+
+provider dnsimple {}
```
Comment (DXTimer) [inline terraform/domains/main.tf:10]: I think creating variables and using them as part of the config would also be a valid approach. In either case we should include in the readme what env variables need to be set for this to work.

**Comment by weppos** [comment]
Not necessary at this time.

**Pull request closed without merging**